### PR TITLE
Refine syntax error recovery for partial code to not interfere with non-error code.

### LINKED ISSUE #17683: Compiler sensitive to newline in parsing an expression
In VS2017 (C# 7), the compiler now rejects the following. Eliminating the newline before `MinValue` allows it to compile. This code compiled in previous versions of the product.

See also our [internal VSO bug](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems?id=393733&_a=edit) tracking this and the [community feedback page](https://developercommunity.visualstudio.com/content/problem/25465/c-linq-expression-doesnt-compile-in-vs2017.html).

``` c#
using System;
using System.Linq;

namespace ConsoleApp1
{
    class Program
    {
        static void Main(string[] args)
        {
            var a = from t in Enumerable.Range(0, 10)
                    where
                    t == Int32.
                    MinValue
                    select t;
        }
    }
}
```

### PR DESCRIPTION
**Pull request by gafter**
Related to #15885
Fixes #17683

@CyrusNajmabadi Would you like to have a look at this before I send it out for review?

### CODE REVIEW

**Comment by tannergooding** [comment]
test windows_debug_vs-integration_prtest please

**Comment by gafter** [comment]
@dotnet/roslyn-compiler @jaredpar @CyrusNajmabadi Please review this tiny correction to a recent change to error recovery. It addresses [VSO 393733](https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_workitems?_a=edit&id=393733&triage=true)

File: src/Compilers/CSharp/Test/Syntax/Parsing/ExpressionParsingTests.cs [hunk 1]
```diff
@@ -3533,5 +3533,127 @@ void M()
...
+        [Fact, WorkItem(17683, "https://github.com/dotnet/roslyn/issues/17683")]
+        public void Bug17683a()
+        {
+            var source =
+@"from t in e
+where
+t == Int32.
+MinValue
+select t";
+            UsingExpression(source);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ExpressionParsingTests.cs]: >UsingExpression(source); [](start = 12, length = 24)

Is there a way to assert that there is no errors?

File: src/Compilers/CSharp/Test/Syntax/Parsing/ExpressionParsingTests.cs [hunk 2]
```diff
@@ -3533,5 +3533,127 @@ void M()
...
+        public void Bug17683b()
+        {
+            var source =
+@"switch (e)
+{
+    case Int32.
+               MaxValue when true:
+            break;
+}";
+            UsingStatement(source);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ExpressionParsingTests.cs]: >UsingStatement(source); [](start = 12, length = 23)

Is there a way to assert that there is no errors?

File: src/Compilers/CSharp/Test/Syntax/Parsing/ExpressionParsingTests.cs (hunk 1 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ExpressionParsingTests.cs]: It looks like the helper is taking expected diagnostics as an optional parameter. So the errors are checked.

File: src/Compilers/CSharp/Test/Syntax/Parsing/ExpressionParsingTests.cs (hunk 2 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ExpressionParsingTests.cs]: It looks like the helper is taking expected diagnostics as an optional parameter. So the errors are checked.

File: src/Compilers/CSharp/Test/Syntax/Parsing/ExpressionParsingTests.cs (hunk 1 above)
Comment (gafter) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ExpressionParsingTests.cs]: The name of the helper doesn't really help the reader know that, though. I'll look at improving that in the future.

**Comment by gafter** [comment]
@dotnet/roslyn-compiler May I please have a second review?

**Comment by gafter** [comment]
@dotnet/roslyn-compiler May I please have a second review?

**Comment by gafter** [comment]
@dotnet/roslyn-compiler May I please have a second review?

**Comment by cston** [comment]
LGTM

**Comment by CyrusNajmabadi** [comment]
👍

**Pull request merged**